README.md
![logo](assets/logo.png)


## Geographical Calculator
[![License](https://poser.pugx.org/kmlaravel/apis-generator/license)](//packagist.org/packages/kmlaravel/laravel-geographical-calculator)
[![PHP Composer](https://github.com/karam-mustafa/laravel-geographical-calculator/actions/workflows/php.yml/badge.svg)](https://github.com/karam-mustafa/laravel-geographical-calculator/actions/workflows/php.yml)
[![Check & fix styling](https://github.com/karam-mustafa/laravel-geographical-calculator/actions/workflows/php-cs-fixer.yml/badge.svg)](https://github.com/karam-mustafa/laravel-geographical-calculator/actions/workflows/php-cs-fixer.yml)
[![Run tests](https://github.com/karam-mustafa/laravel-geographical-calculator/actions/workflows/tests.yml/badge.svg)](https://github.com/karam-mustafa/laravel-geographical-calculator/actions/workflows/tests.yml)
[![Code Quality](https://api.codiga.io/project/30429/score/svg)](https://api.codiga.io/project/30429/score/svg)

Geographical Calculator was developed for laravel 5.8+ to help you to implement geographical calculation, 
with With several algorithms that help you deal with coordinates.

Installation
------------
##### 1 - Dependency
The first step is using composer to install the package and automatically update your composer.json file, you can do this by running:

```shell
composer require kmlaravel/laravel-geographical-calculator
```
##### 2 - Copy the package providers to your local config with the publish command, this will publish the config:
```shell

php artisan geo:install

```

Features
-----------
- [Get the distance between one or more of points](https://github.com/karam-mustafa/laravel-geographical-calculator/blob/main/docs/distances.md#basic-usage)
- [Get the center between set of points](https://github.com/karam-mustafa/laravel-geographical-calculator/blob/main/docs/areas.md#get-the-center-for-a-given-coordinates)
- [Get the closest point to the specific point](https://github.com/karam-mustafa/laravel-geographical-calculator/blob/main/docs/ordering.md#get-closest-point)
- [Get the farthest point to the specific point](https://github.com/karam-mustafa/laravel-geographical-calculator/blob/main/docs/ordering.md#get-farthest-point)
- [Check if a given point located in a given area](https://github.com/karam-mustafa/laravel-geographical-calculator/blob/main/docs/areas.md#check-if-a-given-point-located-in-a-given-area)
- [Get a ranking of points by the nearest neighbor algorithm](https://github.com/karam-mustafa/laravel-geographical-calculator/blob/main/docs/ordering.md#get-ordering-points-by-nearest-neighbor-algorithm)
- [Get all package features](https://github.com/karam-mustafa/laravel-geographical-calculator/blob/main/docs/all.md#all-features-in-one-function)


config options
----------------
> ## add your custom units,keys ..etc.
>
```php
    /*
    |--------------------------------------------------------------------------
    | units values
    |--------------------------------------------------------------------------
    | your custom units, the initial units its convert from a mile to any value
    | (1.609344) is the conversion factor from a mile to a kilometer
    */
    'units' => [
        'mile' => 1,
        'km' => 1.609344,
        'm' => (1.609344 * 1000),
        'cm' => (1.609344 * 100),
        'mm' => (1.609344 * 1000 * 1000),
    ],

    /*
    |--------------------------------------------------------------------------
    | distance_key_prefix
    |--------------------------------------------------------------------------
    | if you declared more than tow points to resolve their distance,
    | you will see the result in the following format:
    | "1-2" => ["km" => "some result"],
    | "2-3" => ["km" => "some result"],
    | "3-4" => ["km" => "some result"],
    | and if you want to set any prefix before each index
    | you must change the below value to any value you want.
    |
    */
    'distance_key_prefix' => '',
```

Changelog
---------
Please see the [CHANGELOG](https://github.com/kmlaravel/laravel-geographical-calculator/blob/master/CHANGELOG.md) for more information about what has changed or updated or added recently.

Security
--------
If you discover any security related issues, please email them first to karam2mustafa@gmail.com, 
if we do not fix it within a short period of time please open a new issue describing your problem. 

Credits
-------
[karam mustafa](https://www.linkedin.com/in/karam2mustafa)


CHANGELOG.md
## v2.2.0 - 2022-02-08

- Added the farthest point from a set of points to a custom point.
- Added check if the point is located in a given area.

## v2.1.1 - 2022-01-22

- Fix code style and composer dependencies

## v2.1.0 - 2021-12-27

- Improved getDistance and getCenter function to accept callback.
- Added all function to get all package features.
- Added Loops trait helper.
- Added the closest point from a set of points to a custom point.
- Added nearest neighbor algorithm.
- Improved tests and code style

## [coming] v2.2.0

- Improved traits design.
- Added check if the point is located in a given area.
- Added the farthest point from a set of points to a custom point.
- Added Features to eloquent model scope.

## v2.1.0

- Improved getDistance and getCenter function to accept callback.
- Added all function to get all package features.
- Added Loops trait helper.
- Added the closest point from a set of points to a custom point.
- Added nearest neighbor algorithm.
- Improved tests and code style

## v2.0.1

- fix if there are no options inserted.

v2.0.0
--------،

- Refactor class names.
- Refactor distances calculations.
- Added documents folder.
- Added helper traits.
- Added install command.
- Added a package logo.
- Added get the center of set of points.
- Use the register pattern to store and fetch the data.

## v1.1.0

- refactor and improve functions definitions
- write a package test


src/Traits/Distances.php
<?php

namespace KMLaravel\GeographicalCalculator\Traits;

use Exception;
use Illuminate\Support\Collection;

trait Distances
{
    /**
     * available units.
     *
     * @author karam mustafa
     *
     * @var array
     */
    private $units = [
        'mile' => 1,
        'km' => 1.609344,
        'm' => (1.609344 * 1000),
        'cm' => (1.609344 * 100),
        'mm' => (1.609344 * 1000 * 1000),
    ];

    /**
     * get the available units.
     *
     * @return array
     *
     * @author karam mustafa
     */
    public function getUnits()
    {
        return $this->units;
    }

    /**
     * @param array $units
     *
     * @return Distances
     *
     * @author karam mustaf
     */
    public function setUnits($units)
    {
        $this->units = $units;

        return $this;
    }

    /**
     * instance constructor.
     *
     * @author karam mustafa
     */
    public function __construct()
    {
        $this->resolveUnits();
    }

    /**
     * Finding the distance of points using several given coordinate points.
     *
     * @param null|callable $callback
     *
     * @return array|\Illuminate\Support\Collection
     *
     * @author karam mustafa
     */
    public function getDistance($callback = null)
    {
        $this->through($this->getPoints(), function ($index, $point) {
            // check if we are not arrive to the last point yet.
            if (isset($this->getPoints()[$index + 1])) {
                // init and calc sin and cos value
                $this->setSin($this->getAngle($point[0], $this->getPoints($index + 1)[0], 'sin'))
                    ->setCos($this->getAngle($point[0], $this->getPoints($index + 1)[0], 'cos'))
                    // set the position of this loop at the local storage.
                    ->setInStorage('position', ($index + 1).'-'.($index + 2))
                    // set first longitude.
                    ->setLongitude($point[1])
                    // set second longitude.
                    ->setLongitude($this->getPoints($index + 1)[1])
                    // set the formatted key that bind with the prefix config.
                    ->setInStorage(
                        'distance_key',
                        $this->formatDistanceKey($this->getFromStorage('position'))
                    )
                    // save the results.
                    ->setResult([$this->getFromStorage('distance_key') => $this->calcDistance()]);
            }
        });

        return $this->resolveCallbackResult($this->cleanDistanceResult(), $callback);
    }

    /**
     * get the sin or cos values multiply.
     *
     * @param int    $firstLat
     * @param int    $secondLat
     * @param string $angle
     *
     * @return float
     *
     * @author karam mustafa
     */
    private function getAngle($firstLat, $secondLat, $angle = 'sin')
    {
        // convert the first value to radian and get result sin or cos method
        // convert the second value to radian and get result sin or cos method
        return $angle(deg2rad($firstLat)) * $angle(deg2rad($secondLat));
    }

    /**
     * get theta angle.
     *
     * @return float
     *
     * @author karam mustafa
     */
    private function getValueForAngleBetween()
    {
        return cos(deg2rad($this->getLongs()[0] - $this->getLongs()[1]));
    }

    /**
     * calculation distance process.
     *
     * @throws Exception
     *
     * @return array
     *
     * @author karam mustafa
     */
    private function calcDistance()
    {
        $this->setInStorage(
            'distance',
            acos($this->getSin() + $this->getCos() * $this->getValueForAngleBetween())
        )->setInStorage(
            'rad2deg',
            rad2deg($this->getFromStorage('distance'))
        )->setInStorage(
            'correctDistanceValue',
            $this->correctDistanceValue($this->getFromStorage('rad2deg'))
        );

        return $this->resolveDistanceWithUnits($this->getFromStorage('correctDistanceValue'));
    }

    /**
     * @param $distance
     *
     * @return float
     *
     * @author karam mustafa
     */
    private function correctDistanceValue($distance)
    {
        return $distance * 60 * 1.1515;
    }

    /**
     * check if user chose any units.
     *
     * @param float $distance
     *
     * @throws Exception
     *
     * @return array
     *
     * @author karam mustafa
     */
    private function resolveDistanceWithUnits($distance)
    {
        if (isset($this->getOptions()['units']) &&
            sizeof($this->getOptions('units')) > 0
        ) {
            // loop in each unit and solve the distance.
            $this->through($this->getOptions()['units'], function ($index, $unit) use ($distance) {
                $this->checkIfUnitExists($unit)
                    // set the result in storage.
                    ->setInStorage($unit, $distance * $this->getUnits()[$unit]);
            });
        } else {
            // if there are not units, then get the default units property.
            $this->setInStorage('mile', $distance * $this->getUnits()['mile']);
        }

        // remove un required results and get the results from storage.
        $this->removeFromStorage('position', 'distance_key', 'distance', 'rad2deg', 'correctDistanceValue');

        return $this->getFromStorage();
    }

    /**
     * check if user chose any units.
     *
     * @author karam mustafa
     */
    private function resolveUnits()
    {
        if (config('geographical_calculator.units')) {
            $this->setUnits(config('geographical_calculator.units'));
        }
    }

    /**
     * check if current units its available in units property or config file.
     *
     * @param string $unit
     *
     * @throws \Exception
     *
     * @return Distances
     *
     * @author karam mustafa
     */
    private function checkIfUnitExists($unit)
    {
        if (! isset($this->getUnits()[$unit])) {
            throw new Exception("the unit ['$unit'] dose not available in units config");
        }

        return $this;
    }

    /**
     * get each point and calculate it distance to the main point.
     *
     * @author karam mustafa
     */
    private function resolveEachDistanceToMainPoint()
    {
        // store the all points in the points key in the storage
        $this->setInStorage('points', $this->getPoints());

        // empty all points after we store them in the storage.
        $this->clearPoints();
        // get the first unit inserted from a config or from the default config
        // the dose not matter what is the unit we use to sort the result.
        $this->setInStorage('unit', collect($this->getUnits())->keys()->first());

        $this->through($this->getFromStorage('points'), function ($index, $point) {

            // calculate distance for each point in the points
            // and append this distance to closestDistance storage key.
            $this->appendToStorage(
                'distancesEachPointToMainPoint',
                $this->setPoints([$this->getMainPoint(), $point])
                    ->getDistance(function (Collection $result) {
                        return $result->first()[$this->getFromStorage('unit')];
                    })
            );

            // re remove the points to calculate a new distance.
            $this->clearPoints();
        });
    }

    /**
     * get only result that related with units.
     *
     * @return mixed
     *
     * @author karam mustafa
     */
    public function cleanDistanceResult()
    {
        return $this->getResult(function ($result) {
            return collect($result)->filter(function ($results) {
                return collect($results)->filter(function ($value, $key) {
                    return in_array($key, array_keys($this->getUnits()));
                });
            });
        })->toArray();
    }
}
